Updated documentation. (#1943)
rack/rack: README.md

# ![Rack](contrib/logo.webp)

> > **_NOTE:_** Rack 3 was recently released. Please check the [Upgrade Guide](UPGRADE-GUIDE.md) for more details about migrating your existing severs, middlewares and applications. For detailed information on specific changes, check the [changelog](CHANGELOG.md).

Reviewer: ```suggestion
> **_NOTE:_** Rack 3 was recently released. Please check the [Upgrade Guide](UPGRADE-GUIDE.md) for more details about migrating your existing servers, middlewares and applications. For detailed information on specific changes, check the [changelog](CHANGELOG.md).
```

> > **_NOTE:_** Rack 3 was recently released. Please check the [Upgrade Guide](UPGRADE-GUIDE.md) for more details about migrating your existing servers, middlewares and applications. For detailed information on specific changes, check the [changelog](CHANGELOG.md).

Author: ```suggestion
> **_NOTE:_** Rack v3.0.0rc1 was recently released. Please check the [Upgrade Guide](UPGRADE-GUIDE.md) for more details about migrating your existing servers, middlewares and applications. For detailed information on specific changes, check the [changelog](CHANGELOG.md).
```

Rack provides a minimal, modular, and adaptable interface for developing
web applications in Ruby. By wrapping HTTP requests and responses in
the simplest way possible, it unifies and distills the bridge between web
servers, web frameworks, and web application into a single method call.

The exact details of this are described in the [Rack Specification],
which all Rack applications should conform to.

## Installation

> Add the rack gem to your application bundle, or follow the instructions provided by a [supported web framework](#supported-web-frameworks):

Author: Okay, done. Feel free to tweak the instructions if you think it can be better.

```bash
$ gem install rackup
$ rackup
$ curl http://localhost:9292
Hello World
```

## Supported web servers

Rack is supported by a wide range of servers, including:

* [Agoo](https://github.com/ohler55/agoo)
* [Falcon](https://github.com/socketry/falcon)
* [Iodine](https://github.com/boazsegev/iodine)
* [NGINX Unit](https://unit.nginx.org/)
* [Phusion Passenger](https://www.phusionpassenger.com/) (which is mod_rack for Apache and for nginx)
* [Puma](https://puma.io/)
* [Thin](https://rubygems.org/gems/thin)
* [Unicorn](https://yhbt.net/unicorn/)
* [uWSGI](https://uwsgi-docs.readthedocs.io/en/latest/)
* [Lamby](https://lamby.custominktech.com) (for AWS Lambda)

You will need to consult the server documentation to find out what features and
limitations they may have. In general, any valid Rack app will run the same on
all these handlers, without changing anything.

### Rackup

Rack provides a separate gem, [rackup](https://github.com/rack/rackup) which is a generic interface for running a Rack application on supported servers, which include `WEBRick`, `Puma`, `Falcon` and others.

## Supported web frameworks

These frameworks and many others support the [Rack Specification]:

> * [Camping](http://www.ruby-camping.com/)

Author: Done.

> * [Camping](http://www.ruby-camping.com/)
> * [Coset](http://leahneukirchen.org/repos/coset/)
> * [Hanami](https://hanamirb.org/)
> * [Padrino](http://padrinorb.com/)
> * [Ramaze](http://ramaze.net/)
> * [Roda](https://github.com/jeremyevans/roda)
> * [Ruby on Rails](https://rubyonrails.org/)
> * [Rum](https://github.com/leahneukirchen/rum)
> * [Sinatra](http://sinatrarb.com/)
> * [Utopia](https://github.com/socketry/utopia)
> * [WABuR](https://github.com/ohler55/wabur)

Reviewer: ```suggestion
* [Camping](http://www.ruby-camping.com/)
* [Coset](https://leahneukirchen.org/repos/coset/)
* [Hanami](https://hanamirb.org/)
* [Padrino](https://padrinorb.com/)
* [Ramaze](http://ramaze.net/)
* [Roda](https://github.com/jeremyevans/roda)
* [Ruby on Rails](https://rubyonrails.org/)
* [Rum](https://github.com/leahneukirchen/rum)
* [Sinatra](https://sinatrarb.com/)
* [Utopia](https://github.com/socketry/utopia)
* [WABuR](https://github.com/ohler55/wabur)
```

## Available middleware shipped with Rack

Between the server and the framework, Rack can be customized to your
applications needs using middleware. Rack itself ships with the following
middleware:

* `Rack::CommonLogger` for creating Apache-style logfiles.

> * `Rack::ConditionalGet` for returning not modified responses when the response

Reviewer: ```suggestion
* `Rack::ConditionalGet` for returning [Not Modified](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/304) responses when the response
```

  has not changed.
* `Rack::Config` for modifying the environment before processing the request.

> * `Rack::ContentLength` for setting content-length header based on body size.
> * `Rack::ContentType` for setting default content-type header for responses.

Reviewer: ```suggestion
* `Rack::ContentLength` for setting a Content-Length header based on body size.
* `Rack::ContentType` for setting a default Content-Type header for responses.
```

I added articles to the sentence, perhaps this makes it flow easier? (Not a native speaker.)

"Header-casing" the mention of specific headers may make them easier to disambiguate from the rest of the text.

Author: Rack 3 requires lower case headers, so it will be confusing, perhaps use backticks?

Reviewer: Wise choice, backticks.

Author: ```suggestion
* `Rack::ContentLength` for setting a `content-length` header based on body size.
* `Rack::ContentType` for setting a default `content-type` header for responses.
```

* `Rack::Head` for returning an empty body for HEAD requests.
* `Rack::Lint` for checking conformance to the [Rack Specification].
* `Rack::Lock` for serializing requests using a mutex.
* `Rack::Logger` for setting a logger to handle logging errors.
* `Rack::MethodOverride` for modifying the request method based on a submitted
  parameter.
* `Rack::Recursive` for including data from other paths in the application,
  and for performing internal redirects.
* `Rack::Reloader` for reloading files if they have been modified.
* `Rack::Runtime` for including a response header with the time taken to
  process the request.
* `Rack::Sendfile` for working with web servers that can use optimized
  file serving for file system paths.
* `Rack::ShowException` for catching unhandled exceptions and
  presenting them in a nice and helpful way with clickable backtrace.
* `Rack::ShowStatus` for using nice error pages for empty client error
  responses.
* `Rack::Static` for more configurable serving of static files.
* `Rack::TempfileReaper` for removing temporary files creating during a
  request.

All these components use the same interface, which is described in

> detail in the Rack specification.  These optional components can be

Author: ```suggestion
detail in the [Rack Specification]. These optional components can be
```

over:

* `Rack::Request` which also provides query string parsing and
  multipart handling.
* `Rack::Response` for convenient generation of HTTP replies and
  cookie handling.
* `Rack::MockRequest` and `Rack::MockResponse` for efficient and quick
  testing of Rack application without real HTTP round-trips.
* `Rack::Cascade` for trying additional Rack applications if an
  application returns a not found or method not supported response.
* `Rack::Directory` for serving files under a given directory, with
  directory indexes.
* `Rack::MediaType` for parsing content-type headers.
* `Rack::Mime` for determining content-type based on file extension.
* `Rack::RewindableInput` for making any IO object rewindable, using
  a temporary file buffer.
* `Rack::URLMap` to route to multiple applications inside the same process.

## Configuration

Rack exposes several configuration parameters to control various features of
the implementation.

### `param_depth_limit`

```ruby
Rack::Utils.param_depth_limit = 32 # default
```

The maximum amount of nesting allowed in parameters.
For example, if set to 3, this query string would be allowed:

```

> ?a](b]](c]=d

Reviewer 2: This should probably be `?a[b][c]=d`

Author: ```suggestion
?a[b][c]=d
```

```

but this query string would not be allowed:

```

> ?a](b]](c]](d]=e

Reviewer 2: This should probably be `?a[b][c][d]=e`

Author: ```suggestion
?a[b][c][d]=e
```

Sorry, I totally messed up the formatting of this by an unrelated find and replace that went wrong.

```

Limiting the depth prevents a possible stack overflow when parsing parameters.

### `multipart_part_limit`

```ruby
Rack::Utils.multipart_part_limit = 128 # default
```

The maximum number of parts a request can contain.

> Accepting too many part can lead to the server running out of file handles.

Reviewer: ```suggestion
Accepting too many parts can lead to the server running out of file handles.
```

The default is 128, which means that a single request can't upload more than
128 files at once. Set to 0 for no limit.

Can also be set via the `RACK_MULTIPART_PART_LIMIT` environment variable.

## Changelog

See [CHANGELOG.md](CHANGELOG.md).

## Contributing

See [CONTRIBUTING.md](CONTRIBUTING.md) for specific details about how to make a
contribution to Rack.

Please post bugs, suggestions and patches to
[GitHub Issues](https://github.com/rack/rack/issues).

Please check our [Security Policy](https://github.com/rack/rack/security/policy)
for responsible disclosure and security bug reporting process. Due to wide
usage of the library, it is strongly preferred that we manage timing in order
to provide viable patches at the time of disclosure. Your assistance in this
matter is greatly appreciated.

## See Also

### `rack-contrib`

The plethora of useful middleware created the need for a project that
collects fresh Rack middleware. `rack-contrib` includes a variety of
add-on components for Rack and it is easy to contribute new modules.

* https://github.com/rack/rack-contrib

### `rack-session`

> Provides convenient server side session management for Rack.

Reviewer: ```suggestion
Provides convenient server side-session management for Rack.
```

Hyphens, they're so useful.

Author: I believe this should actually be `server-side session`?

Reviewer: haha, YES.

Reviewer 2: I would remove `server side` completely, because it gives the impression that sessions are also stored server side.  The most common session middleware is the cookie one, in which sessions are stored client side.

Author: ```suggestion
Provides convenient session management for Rack.
```

* [Hanami](https://hanamirb.org/)
* [Padrino](https://padrinorb.com/)
* [Roda](https://github.com/jeremyevans/roda) **(Rack 3 Compatible)**
* [Ruby on Rails](https://rubyonrails.org/)
* [Sinatra](https://sinatrarb.com/)
* [Utopia](https://github.com/socketry/utopia) **(Rack 3 Compatible)**
* [WABuR](https://github.com/ohler55/wabur)

### Older (possibly unsupported) web frameworks

* [Ramaze](http://ramaze.net/)
* [Rum](https://github.com/leahneukirchen/rum)

## Available middleware shipped with Rack

Between the server and the framework, Rack can be customized to your
applications needs using middleware. Rack itself ships with the following
middleware:

* `Rack::CommonLogger` for creating Apache-style logfiles.
* `Rack::ConditionalGet` for returning [Not
  Modified](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/304)
  responses when the response has not changed.
* `Rack::Config` for modifying the environment before processing the request.
* `Rack::ContentLength` for setting a `content-length` header based on body
  size.
* `Rack::ContentType` for setting a default `content-type` header for responses.
* `Rack::Deflater` for compressing responses with gzip.

> * `Rack::ETag` for setting `etag` header on string bodies.

Reviewer 2: `s/string bodies/array bodies/` (I think).

Author: ```suggestion
* `Rack::ETag` for setting `etag` header on bodies that can be buffered.
```
